Board Speed: Implemented pushback and haptic feedback

buestad opened the pull request:
Feature: Added functionality for speed-based pushback and haptic buzz. Configurable in Refloat Cfg

https://github.com/lukash/refloat/issues/46

lukash commented:
Thanks for the PR, skimmed through, looks pretty clean. I'll have to think about the new config options (why not use the same settings as for duty pushback?) and the new pushback SAT (here a specific SAT might be the way to go, but in general I wanna make one SAT be "alert pushback" and then have another enum as the alert reason).

The CI has failed on `clang-format`, there's a documentation on how to run it in the cotribution guidelines. I'll be busy the next week and a bit, might be slow to response...

buestad commented:
I Tried running the clang-format (also throug lefthook) but it fails:

```
$ clang-format -i src/main.c 
.../refloat/.clang-format:7:24: error: invalid boolean
AlignTrailingComments: Never
                       ^~~~~
```

buestad commented:
About the config options, It might make sense to use the same angle and angle-speed as duty, but its nice to have a different haptic tone. I don't wanna ignore duty, but speed, not so important :)

lukash commented:
> I Tried running the clang-format (also throug lefthook) but it fails:
> 
> ```
> $ clang-format -i src/main.c 
> .../refloat/.clang-format:7:24: error: invalid boolean
> AlignTrailingComments: Never
>                        ^~~~~
> ```

Looks like you have an old version of clang-format, you need version 18.

buestad commented:
> Looks like you have an old version of clang-format, you need version 18.

Fixed

lukash commented:
@buestad can you please squash the clang-format fix with the main commit and also drop the merge commits? Instead (if needed) please _rebase_ the branch onto the current `main`.

lukash commented:
Also, can you just globally replace `board_speed` for just `speed` and analogously for other ways of writing it (SCREAMING_CASE, CamelCase and whatnot) :grin:? The word "board" is redundant...

And, in general the goal is to keep the config options minimal necessary and I think we can use the duty settings for mostly everything, the only new option should be the actual speed setting. The options for duty cycle which will now also be used for speed alerting just need to be updated (names and descriptions) to say they now apply to Speed too.

lukash commented on src/main.c:
```diff
@@ -644,6 +648,10 @@ static void calculate_setpoint_target(Data *d) {
             d->state.sat = SAT_NONE;
             d->setpoint_target = 0;
         }
+    } else if (d->float_conf.tiltback_board_speed > 0.0 && d->motor.speed > d->float_conf.tiltback_board_speed) {
```
The `d->motor.speed > d->float_conf.tiltback_board_speed` part of the condition should take absolute value to apply also when going battery forward.

lukash commented on src/main.c:
```diff
@@ -644,6 +648,10 @@ static void calculate_setpoint_target(Data *d) {
         }
+    } else if (d->float_conf.tiltback_board_speed > 0.0 && d->motor.speed > d->float_conf.tiltback_board_speed) {
+        d->beep_reason = BEEP_BOARD_SPEED;
+        d->setpoint_target = d->float_conf.tiltback_board_speed_angle;
```
... and here you need to do the same thing as in the other branches, depending on the direction make the setpoint target negative, to work also when going battery forward.

buestad commented:
I can do the fixes tomorrow 👍 About board_speed. The reason for choosing that was because some other variables in the same area that had speed in the name. Believe it was related to the speed of pushback or something.

lukash commented on src/conf/datatypes.h:
```diff
@@ -218,6 +218,7 @@ typedef struct {
     float tiltback_lv_angle;
     float tiltback_lv_speed;
     float tiltback_lv;
+    float tiltback_speed;
```
Nitpick, can you please move this below `tiltback_duty` a few lines above, just so that the related options are together?

buestad replied:
fixed

lukash commented on src/conf/settings.xml:
```diff
@@ -733,6 +733,32 @@ p, li { white-space: pre-wrap; }
             <vTx>7</vTx>
         </tiltback_lv>
+        <tiltback_speed>
+            <longName>Board Speed Limit</longName>
```
"Speed Threshold" to keep it consistent with "Duty Cycle Threshold".

buestad replied:
fixed

lukash commented on src/conf/settings.xml:
```diff
@@ -733,6 +733,32 @@ p, li { white-space: pre-wrap; }
+            <vTxDoubleScale>1000</vTxDoubleScale>
+            <suffix> km/h</suffix>
+            <vTx>7</vTx>
+        </tiltback_speed>
```
This option can be of type Int, TX Type UINT_8, that saves on the total config size. And the step would be better at 5.

At a later point I'll edit the option description a bit myself, it'll be easier than commenting on the bloated HTML text...

buestad replied:
fixed

lukash replied:
The option is still of type Double and step is still 1...

buestad replied:
Oh. I misundestood. I thaugt that vTx = 5 ment double. Should vTx be 5 or 7?  What is vTx? You think that step size in km/h should be 5? Isn't that a bit course?

buestad replied:
fixed. Used VESC param editor to review and change the XML

lukash commented on src/conf/settings.xml:
```diff
@@ -871,7 +897,7 @@ p, li { white-space: pre-wrap; }
 p, li { white-space: pre-wrap; }
 &lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:'Roboto'; ; font-weight:400; font-style:normal;&quot;&gt;
-&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Audible frequency of the Duty Cycle haptic feedback. Also used for Current Limit haptic Feedback, if enabled.&lt;/p&gt;
+&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;Audible frequency of the Duty Cycle haptic feedback. Also used for Current Limit and Board Speed haptic Feedback, if enabled.&lt;/p&gt;
```
Instead of "Board Speed" I'd use "Speed alert" here, ditto on the strength right below.

buestad replied:
fixed

lukash commented on src/haptic_feedback.c:
```diff
@@ -83,6 +85,8 @@ static uint8_t get_beats(HapticFeedbackType type) {
     case HAPTIC_FEEDBACK_DUTY_CONTINUOUS:
         return 0;
+    case HAPTIC_FEEDBACK_SPEED:
+        return 3;
```
I'm not sure off the bat what tone pattern this does, but here I'd keep it simple and use the same pattern as duty. That means you don't need the extra `HAPTIC_FEEDBACK_SPEED`, you can reuse `HAPTIC_FEEDBACK_DUTY` and ideally rename it to something like `HAPTIC_FEEDBACK_SPEED_DUTY`.

buestad replied:
Decided to keep `HAPTIC_FEEDBACK_SPEED`. but tidy up the cases in stead. Felt kinda messy to use a common one because of the `HAPTIC_FEEDBACK_DUTY_CONTINUOUS`

lukash replied:
But it's now redundant, it represents the same thing as `HAPTIC_FEEDBACK_DUTY`, please unify it into one item...

buestad replied:
Changed `HAPTIC_FEEDBACK_DUTY` to `HAPTIC_FEEDBACK_DEFAULT` and adopted

lukash commented on src/conf/settings.xml:
```diff
@@ -3968,6 +3995,8 @@ p, li { white-space: pre-wrap; }
                     <param>tiltback_duty_angle</param>
                     <param>tiltback_duty_speed</param>
                     <param>is_dutybeep_enabled</param>
+                    <param>::sep::Board Speed Alert Pushback</param>
```
"Speed Alert Pushback" here I guess. As I said the options need some restructuring, there's not much point to have a section with only one item, ...

buestad replied:
fixed

lukash commented on src/haptic_feedback.c:
```diff
@@ -55,8 +55,10 @@ static HapticFeedbackType haptic_feedback_get_type(
             return HAPTIC_FEEDBACK_DUTY_CONTINUOUS;
         } else {
-            return HAPTIC_FEEDBACK_DUTY;
+            return HAPTIC_FEEDBACK_DEFAULT;
```
While I don't think `HAPTIC_FEEDBACK_SPEED_DUTY` is a great name, I think it's still better than `HAPTIC_FEEDBACK_DEFAULT`, which carries no information about what it represents... Could you use that? (or provide arguments why yours is better :grin:)

buestad replied:
I think that default is better because it is general. Now this goes for two or three things (duty, speed and maybe current?) But I think maybe we can do a refactor later if needed. Changed for now to `HAPTIC_FEEDBACK_SPEED_DUTY ` and we can discuss in Discord :)

lukash replied:
The current is using `DUTY_CONTINUOUS` (not reflected in the name) :sweat_smile:

buestad committed:
Add support for speed-based alerting

Feature: Add support for speed-based alerting (Pushback and Haptic)
 >
 Trigger Pushback and Haptic Feedback at a desired speed. The properties
 other than the Speed Threshold are shared with Duty Cycle.

lukash commented:
I changed the target branch for the PR from `main` to `testing`, we normally do this for features that should get some testing (and I move them to `main` manually later on).

lukash committed:
Reorganize Pushback Alerts and Haptic Feedback config options

Take the new Speed Pushback into account and clean up some descriptions.

acheronfail commented:
Just wanted to stop by and mention I’ve been running this with a speed limit of 15kmh on a PintV that my son’s been using for 2 weeks and have had no issues whatsoever so far 👍
